### PR TITLE
Bug 1707891 - Implement shutdown API

### PR DESCRIPTION
**Pull request by brizental**
The lack of this API causes a bug in Rally. See: https://bugzilla.mozilla.org/show_bug.cgi?id=1698934#c33

#### Note

I have confirmed that this fixes the issue and makes the test over on https://github.com/mozilla-rally/rally-core-addon/pull/678 pass.

### Pull Request checklist ###

- [x] **Quality**: Make sure this PR builds and runs cleanly.
  - Inside the `glean/` folder, run:
    - `npm run test` Runs _all_ tests
    - `npm run lint && npm run lint:circular-deps` Runs _all_ linters
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
- [x] **Documentation**: This PR includes documentation changes, an explanation of why it does not need that or a follow-up bug has been filed to do that work
   - https://github.com/mozilla/glean/pull/1722

### CODE REVIEW

File: glean/src/core/dispatcher.ts [hunk 1]
```diff
@@ -31,6 +35,8 @@ const enum Commands {
   Stop,
   // The dispatcher should stop executing the queued tasks and clear the queue.
   Clear,
+  // The dispatcher will clear the queue and go on Shutdown state.
```
Comment (travis79) [inline glean/src/core/dispatcher.ts:38]:
```suggestion
  // The dispatcher will clear the queue and go into the Shutdown state.
```

File: glean/src/core/dispatcher.ts [hunk 2]
```diff
@@ -286,6 +307,27 @@ class Dispatcher {
     }
   }
 
+  /**
+   * Shutsdown the dispatcher.
+   *
+   * 1. Executes all tasks launched prior to this one.
+   * 2. Clears the queue of any tasks launched after this one.
+   * 3. Puts the dispatcher in a `Shutdown` stated.
```
Comment (travis79) [inline glean/src/core/dispatcher.ts:315]:
```suggestion
   * 3. Puts the dispatcher in the `Shutdown` state.
```

File: glean/src/core/dispatcher.ts [hunk 3]
```diff
@@ -286,6 +307,27 @@ class Dispatcher {
...
+   *
+   * # Note
+   *
+   * - This is a command like any other, if the dispatcher is uninitialized
+   *   it will get executed when the dispatcher is initialized.
+   * - If the dispatcher is stopped it is resumed and all pending tasks are executed.
```
Comment (travis79) [inline glean/src/core/dispatcher.ts:321]:
```suggestion
   * - If the dispatcher is stopped, it is resumed and all pending tasks are executed.
```

File: glean/src/core/glean.ts [hunk 4]
```diff
@@ -418,6 +413,27 @@ class Glean {
     });
   }
 
+  /**
+   * Finishes executing all pending tasks
+   * and shutsdown both Glean's dispatcher and ping uploader.
```
Comment (travis79) [inline glean/src/core/glean.ts:418]:
```suggestion
   * and shuts down both Glean's dispatcher and the ping uploader.
```

File: glean/src/core/glean.ts [hunk 5]
```diff
@@ -418,6 +413,27 @@ class Glean {
...
+   *
+   * This is irreversible.
+   * Only a restart will return Glean back to an idle state.
+   *
+   * @returns A promise which resolves once the shutdown is complete.
+   */
+  static async shutdown(): Promise<void> {
+    // Order here matters!
+    //
+    // The main dispatcher needs to be shutdown first,
```
Comment (travis79) [inline glean/src/core/glean.ts:430]: Native English speaker uber-nit: "shutdown" works fine as a noun, but it is probably more correct grammar to say "shut down" when using it as a verb.

File: glean/src/core/upload/index.ts [hunk 6]
```diff
@@ -68,17 +71,15 @@ class PingBodyOverflowError extends Error {
...
-  // This is `undefined` in case there is no current uploading job.
-  private currentJob?: Promise<void>;
+  // A list of pings currently being processed.
+  private processing: QueuedPing[];
+  // Local dispathcer instance to handle execution of ping requests.
+  private dispatcher: Dispatcher;
   // The object that concretely handles the ping transmission.
   private readonly uploader: Uploader;
-
+  // PlatfornInfo object containing OS information used to build ping request headers.
```
Comment (travis79) [inline glean/src/core/upload/index.ts:80]:
```suggestion
  // PlatformInfo object containing OS information used to build ping request headers.
```

File: glean/src/core/upload/index.ts [hunk 7]
```diff
@@ -106,22 +109,36 @@ class PingUploader implements PingsDatabaseObserver {
    */
   private enqueuePing(ping: QueuedPing): void {
     let isDuplicate = false;
-    for (const queuedPing of this.queue) {
+    for (const queuedPing of this.processing) {
       if (queuedPing.identifier === ping.identifier) {
         isDuplicate = true;
```
Comment (travis79) [inline glean/src/core/upload/index.ts:114]: Once you know this is a duplicate, you could bail out of the for/each loop or just `return;`

File: glean/src/core/upload/index.ts [hunk 8]
```diff
@@ -106,22 +109,36 @@ class PingUploader implements PingsDatabaseObserver {
...
-  }
-
-  /**
-   * Gets and removes from the queue the next ping in line to be uploaded.
-   *
-   * @returns The next ping or `undefined` if no pings are enqueued.
-   */
-  private getNextPing(): QueuedPing | undefined {
-    return this.queue.shift();
+    if (!isDuplicate) {
```
Comment (travis79) [inline glean/src/core/upload/index.ts:118]: I'm not sure what the overhead of variable declaration and instantiation is in JS, but the `isDuplicate` could be avoided if you just returned when a duplicate was detected in the for/each above.

File: glean/src/core/upload/index.ts [hunk 9]
```diff
@@ -216,6 +233,15 @@ class PingUploader implements PingsDatabaseObserver {
     }
   }
 
+  /**
+   * Removes a ping from the processing list.
+   *
+   * @param identifier The identifier of the ping to be removed.
+   */
+  private concludePingProcessing(identifier: string): void {
+    this.processing = this.processing.filter(ping => ping.identifier !== identifier);
```
Comment (travis79) [inline glean/src/core/upload/index.ts:242]: `filter` seems like overkill to remove a single element, but I guess it also has the advantage of removing duplicates in the case that they might exist 🤔

File: glean/src/core/upload/index.ts [hunk 10]
```diff
@@ -281,95 +310,52 @@ class PingUploader implements PingsDatabaseObserver {
...
-      this.status = PingUploaderStatus.Idle;
-    }
+  update(identifier: string, ping: PingInternalRepresentation): void {
+    this.dispatcher.resume();
+    this.enqueuePing({ identifier, retries: 0, ...ping });
   }
 
   /**
-   * Cancels ongoing upload.
+   * Shutsdown internal dispatcher, after executing all previously enqueued ping requests.
```
Comment (travis79) [inline glean/src/core/upload/index.ts:327]:
```suggestion
   * Shuts down internal dispatcher, after executing all previously enqueued ping requests.
```

File: glean/src/core/upload/index.ts [hunk 11]
```diff
@@ -281,95 +310,52 @@ class PingUploader implements PingsDatabaseObserver {
...
-  async clearPendingPingsQueue(): Promise<void> {
-    await this.cancelUpload();
-    this.queue = [];
+  clearPendingPingsQueue(): void {
+    this.dispatcher.clear();
+    this.processing = [];
+    // Create and initialize a new dispatcher so we don't need to wait
+    // on the previous one finishing execution.
+    //
+    // It will only finish execution of the current task, all other queues tasks are dropped.
```
Comment (travis79) [inline glean/src/core/upload/index.ts:346]:
```suggestion
    // It will only finish execution of the current task, all other queued tasks are dropped.
```

File: glean/src/index/qt.js [hunk 12]
```diff
@@ -91,4 +91,15 @@ function setSourceTags(value) {
   Glean.Glean.default.setSourceTags(value);
 }
 
+/**
+ * Finishes executing any ongoing tasks and shutsdown Glean.
```
Comment (travis79) [inline glean/src/index/qt.js:95]:
```suggestion
 * Finishes executing any ongoing tasks and shuts down Glean.
```

File: glean/src/index/qt.ts [hunk 13]
```diff
@@ -101,6 +101,17 @@ export default {
     Glean.setSourceTags(value);
   },
 
+  /**
+   * Finishes executing any ongoing tasks and shutsdown Glean.
```
Comment (travis79) [inline glean/src/index/qt.ts:105]:
```suggestion
   * Finishes executing any ongoing tasks and shuts down Glean.
```

File: glean/src/index/webext.ts [hunk 14]
```diff
@@ -73,6 +73,17 @@ export default {
     Glean.setDebugViewTag(value);
   },
 
+  /**
+   * Finishes executing any ongoing tasks and shutsdown Glean.
```
Comment (travis79) [inline glean/src/index/webext.ts:77]:
```suggestion
   * Finishes executing any ongoing tasks and shuts down Glean.
```

File: glean/tests/unit/core/glean.spec.ts [hunk 15]
```diff
@@ -504,4 +505,25 @@ describe("Glean", function() {
...
+    await Glean.shutdown();
+    assert.strictEqual(Context.dispatcher["state"], DispatcherState.Shutdown);
+    assert.strictEqual(Glean["pingUploader"]["dispatcher"]["state"], DispatcherState.Shutdown);
+  });
+
+  it("shutdown allows all pending pings to be sent before shutting down uploader", async function() {
+    const postSpy = sandbox.spy(Glean.platform.uploader, "post");
+
+    // `setUploadEnabled` is a task dispatched on the main dispatcher,
+    // inside this task an uploading task is dispatched to the ping uploader dispatcher.
```
Comment (travis79) [inline glean/tests/unit/core/glean.spec.ts:519]: This sentence is a little confusing, maybe reword it to something like "this task will dispatch an upload task to the upload dispatcher"

**Review by travis79 — APPROVED** [review]
This looks really good to me!  Mostly just some grammar and clarification nits.

File: glean/src/core/upload/index.ts (hunk 9 above)
Comment (brizental) [reply glean/src/core/upload/index.ts:242]: Yeah, I used filter just because the syntax is more succint. The other way to do this would be something like:

```js
const index = array.indexOf(identifier);
if (index > -1) {
  this.processing.splice(index, 1);
}
```

File: glean/src/core/glean.ts (hunk 5 above)
Comment (brizental) [reply glean/src/core/glean.ts:430]:
```suggestion
    // The main dispatcher needs to be shut down first,
```

File: glean/tests/unit/core/glean.spec.ts (hunk 15 above)
Comment (brizental) [reply glean/tests/unit/core/glean.spec.ts:519]:
```suggestion
    // this task will dispatch an upload task to the upload dispatcher.
```
